### PR TITLE
`Exam mode`: Add warning message for visibility date of an exam

### PR DESCRIPTION
**Pull request by edkaya**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This feature is a part of the new approach for exam notifications discussed in #9121. Exams are not supposed to be visible multiple days before. Typically, they are visible a few hours before the exam starts when all information is correct. We should display a warning when instructors set the visible date too early, which is more than 4h difference.

### Description

Added a warning message in case of 4h difference. 

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to Exams
3. Click Create Exam
4. Set the difference between visible and start date of the exam more than 4h
5. Verify that a warning message is displayed below date picker
6. Verify that tooltip is displayed on hover

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

Screenshot 2024-07-28 at 18 21 39

## Summary by CodeRabbit

- **New Features**
	- Added enhanced visibility management for exam dates, providing users with warnings when visibility dates are set too early.
	- Implemented a new warning feedback mechanism in the date-time picker component for better user guidance.
	- Introduced new warning messages in both English and German for localized user feedback regarding date settings.

- **Bug Fixes**
	- Improved logic to prevent scheduling conflicts due to early exam visibility dates.

- **Tests**
	- Added comprehensive tests for the exam visibility logic to ensure proper functionality based on date conditions.

### CODE REVIEW

**Review by BBesrour — DISMISSED** [review]
Tested on TS6, works as described.

Could you change the warning to `The 'Visible from' date is set too early. We recommend setting it to at most 4 hours before the start of the exam.`. I think it would be more clear.

**Review by anian03 — APPROVED** [review]
Tested on TS6, works as expected.
Agreeing with the previous comment, that makes the warning easier to understand.

I'd also consider shortening the tooltip (i.e. leaving out the last bit explaining that things are up to date, it seems kind of self explanatory in that context)

**Review by iyannsch — DISMISSED** [review]
Code LGTM 👍
Regarding the wording, I think `"warningError": "'Visible from' is set earlier than recommended, more than 4 hours from 'start of working time'"` is not well understandable due to the comma construct. How about "'Visible from' is set more than 4h earlier than 'start of working time'. This is not recommended." or smt like this? 

Also See @BBesrour 's comment :)

**Review by kaancayli — DISMISSED** [review]
Code looks good 👍🏻

File: src/main/webapp/app/shared/date-time-picker/date-time-picker.component.html [hunk 1]
```diff
@@ -43,3 +43,6 @@
 @if (dateInput.invalid || (requiredField && !dateInput.value)) {
     <div class="invalid-feedback">{{ 'entity.dateMissingOrNotValid' | artemisTranslate: { labelName: labelName } }}</div>
 }
+@if (warning) {
+    <div class="invalid-feedback warning" jhiTranslate="entity.warningError" ngbTooltip="{{ 'entity.warningToolTip' | artemisTranslate }}"></div>
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/date-time-picker/date-time-picker.component.html:47]: You can use `text-warning`, so you don't need to create an extra `warning` class. However, displaying yellow text on a white background is hardly readable. We should consider other ways to display the warning message, such as displaying a yellow warning triangle and keeping the text in black.

**Review by rabeatwork — CHANGES_REQUESTED** [review]

**Review by kaancayli — DISMISSED** [review]
Reapprove

**Review by anian03 — APPROVED** [review]
Just re-tested it on TS4, everything works as expected and feedback was addressed :)

**Review by iyannsch — APPROVED** [review]
Code LGTM

**Review by rabeatwork — APPROVED** [review]
Tested on TS5. Thanks for incorporating the feedback.

**Review by kaancayli — APPROVED** [review]

**Review by anian03 — APPROVED** [review]

**Pull request merged**